### PR TITLE
Move everything into src/ocean_emulators, get rid of ocean_emulators_main

### LINKED ISSUE #75: Remove ocean_emulators_main module
Remove this module since we do not depend on it anymore. If we do require some functionality (like [xarray-schema](https://github.com/m2lines/ocean_emulators/blob/main/ocean_emulators/schema.py)), we can port the code over to this repository.

### PR DESCRIPTION
**Pull request by jder**
Follow-up to https://github.com/suryadheeshjith/Ocean_Emulator/pull/108 which moves everything into a named module and fixes #75.

### CODE REVIEW

**Comment by jder** [comment]
This is more-or-less ready to review. I'm working on ignoring and/or fixing the now-working mypy errors so will ping again once I've done that. There are about 30 more to go. :)

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -6,26 +6,26 @@ repos:
...
+# https://whynothugo.nl/journal/2023/01/12/notes-on-pre-commit/#using-the-system-plugin-language
+- repo: local
+  hooks:
+  - id: mypy
+    name: mypy
+    language: system
+    require_serial: true
+    entry: uv run --dev mypy
+    types_or: [python, pyi]
+    args: ["--config", "pyproject.toml"]
```
Comment (jder) [inline .pre-commit-config.yaml:28]: Now that we have new import paths, we can get mypy working. This does that.

File: pyproject.toml [hunk 2]
```diff
@@ -85,5 +87,8 @@ testpaths = [ "tests" ]
 addopts = "--jaxtyping-packages=data,datasets,beartype.beartype"
 
 [tool.mypy]
-ignore_missing_imports = true
 check_untyped_defs = true
+
+[[tool.mypy.overrides]]
+module = [ "cftime.*" ]
+follow_untyped_imports = true
```
Comment (jder) [inline pyproject.toml:94]: cftime doesn't have type information but only exposes a few functions, so we tell mypy to just do its best.

File: src/ocean_emulators/aggregator/inference/main.py [hunk 3]
```diff
@@ -4,8 +4,8 @@
 import wandb
 import xarray as xr
 
-from utils.data import Normalize, get_norm_unnorm_dicts
-from utils.model import InfOutput, SingleTimeseriesOutput
+from ocean_emulators.utils.data import Normalize, get_norm_unnorm_dicts
+from ocean_emulators.utils.model import InfOutput, SingleTimeseriesOutput
```
Comment (jder) [inline src/ocean_emulators/aggregator/inference/main.py:8]: I used `ocean_emulators...` for import paths rather than relative imports, because my IDE (and I assume most IDEs) use that by default, but open to feedback. I didn't go through and fix existing relative paths, maybe we can do that over time.

File: src/ocean_emulators/aggregator/train.py [hunk 4]
```diff
@@ -44,7 +44,8 @@ def get_logs(self, label: str = "train") -> Dict[str, torch.Tensor]:
             **var_loss_dict,
             **channel_loss_dict,
         }
+        meaned_logs = {}
         for key in sorted(logs.keys()):
-            logs[key] = float(all_reduce_mean(logs[key].detach()).cpu().numpy())
+            meaned_logs[key] = float(all_reduce_mean(logs[key].detach()).cpu().numpy())
 
-        return logs
+        return meaned_logs
```
Comment (jder) [inline src/ocean_emulators/aggregator/train.py:50]: I'm not actually sure the right semantics here -- doing this causes errors downstream which makes it unclear to me how this was supposed to work. @suryadheeshjith do you know what the expected types here are?

Comment (suryadheeshjith) [reply src/ocean_emulators/aggregator/train.py:50]: ~~Why are you creating a new dict meaned_logs?~~ seems fine. It looks right, the returned dictionary seems correctly typed as Dict[str, float]

Comment (jder) [reply src/ocean_emulators/aggregator/train.py:50]: The type of `logs` is `Dict[str, Tensor]` but then here we are storing floats, not tensors. So created a new dict to return which is of type `Dict[str, float]` and updated the return type. Not sure if that's the right fix, though.

Comment (suryadheeshjith) [reply src/ocean_emulators/aggregator/train.py:50]: What error are you facing right now?

Comment (jder) [reply src/ocean_emulators/aggregator/train.py:50]: https://github.com/suryadheeshjith/Ocean_Emulator/actions/runs/13956359284/job/39068325637?pr=110 is all of them, the main one I'm not sure about is:

```
src/ocean_emulators/aggregator/validate/main.py:78: error: Signature of "get_logs" incompatible with supertype "TrainAggregator"  [override]
src/ocean_emulators/aggregator/validate/main.py:78: note:      Superclass:
src/ocean_emulators/aggregator/validate/main.py:78: note:          def get_logs(self, label: str = ...) -> dict[str, float]
src/ocean_emulators/aggregator/validate/main.py:78: note:      Subclass:
src/ocean_emulators/aggregator/validate/main.py:78: note:          def get_logs(self, label: str) -> Any
```

Comment (jder) [reply src/ocean_emulators/aggregator/train.py:50]: ie not sure what the type of get_logs should be

Comment (suryadheeshjith) [reply src/ocean_emulators/aggregator/train.py:50]: The error is coming from validate/main.py, try type hinting get_logs() there

Comment (jder) [reply src/ocean_emulators/aggregator/train.py:50]: Thanks, yeah, I guess I'm trying to ask: do you know what the type of `get_logs` should be? If`Dict[str, float]` makes sense then I can carry that throughout.

Comment (suryadheeshjith) [reply src/ocean_emulators/aggregator/train.py:50]: Oh yeah, it should be Dict[str, float].

Comment (suryadheeshjith) [reply src/ocean_emulators/aggregator/train.py:50]: Rechecking the code, the get_logs() function of the ValidateAggregator can return a dict that has values that need not be float. If it is causing problems, we can ignore types for now and find a better solution (maybe decouple the train and validation aggregators) in a later PR.

Comment (suryadheeshjith) [reply src/ocean_emulators/aggregator/train.py:50]: I believe the get_logs() return type in the ValidationAggregator class should be something like Dict[str, float | wandb.Image]]

File: .github/workflows/pre-commit.yml [hunk 5]
```diff
@@ -15,6 +15,9 @@ jobs:
       - name: Check out code
         uses: actions/checkout@v2
 
+      - name: Install uv
+        uses: astral-sh/setup-uv@v5
+
```
Comment (jder) [inline .github/workflows/pre-commit.yml:20]: This is now needed for pre-commit to run.

File: .vscode/settings.json [hunk 6]
```diff
@@ -0,0 +1,5 @@
+{
+    "mypy-type-checker.args": [
+        "\"--config=pyproject.toml\""
+    ]
+}
```
Comment (jder) [inline .vscode/settings.json]: Not sure what we want our policy to be about checking in workspace settings files. If you're using the mypy extension in vscode, you need this to make it work correctly.

File: src/ocean_emulators/aggregator/inference/main.py [hunk 7]
```diff
@@ -238,11 +239,7 @@ def __init__(
             metadata=metadata,
         )
         self._aggregators = aggregators
-        self._summary_aggregators = {
-            name: aggregators[name]
-            for name in ["time_mean"]  # ["time_mean", "annual"]
-        }
-        self._time_dependent_aggregator_names: list[str] = []  # ["annual"]
```
Comment (jder) [inline src/ocean_emulators/aggregator/inference/main.py:245]: `_time_dependent_aggregator_names` appeared to be unused and to be complex from a typing POV below, so I removed it. @suryadheeshjith that seem ok?

File: src/ocean_emulators/aggregator/validate/main.py [hunk 8]
```diff
@@ -34,8 +36,15 @@ def __init__(
...
+    # TODO(jder): we could remove this by moving from inheritance
+    # to composition with the TrainAggregator functionality.
+    def record_batch(self, batch):
+        raise NotImplementedError(
+            "Call record_validation_batch instead of record_batch"
+        )
+
     @torch.no_grad()
-    def record_batch(self, batch: ValOutput):
+    def record_validation_batch(self, batch: ValOutput):
```
Comment (jder) [inline src/ocean_emulators/aggregator/validate/main.py:47]: You can't override record_batch taking ValOutput since the super class says it must take TrainOutput. So I broke out a separate method you should call instead.

File: src/ocean_emulators/aggregator/validate/main.py [hunk 9]
```diff
@@ -104,7 +116,7 @@ def _get_loss_scaled_mse_components(
             rmse_key = f"{label}/mean/weighted_rmse/{var}"
             if rmse_key in validation_metrics:
                 scaled_squared_errors[var] = (
-                    validation_metrics[rmse_key] / self._loss_scaling[var].item()
+                    validation_metrics[rmse_key] / self._loss_scaling[var]
```
Comment (jder) [inline src/ocean_emulators/aggregator/validate/main.py:119]: I think this was an actual bug -- this dict appears to have floats in it, not tensors.

File: src/ocean_emulators/aggregator/validate/map.py [hunk 10]
```diff
@@ -38,7 +38,7 @@ def __init__(self, metadata: Dict[str, Dict[str, str]] = {}, hist: int = 0):
     @torch.no_grad()
     def record_batch(
         self,
-        loss: float,
+        loss,
```
Comment (jder) [inline src/ocean_emulators/aggregator/validate/map.py:41]: This was actually a tensor, I think.

File: src/ocean_emulators/aggregator/validate/reduced.py [hunk 11]
```diff
@@ -100,13 +101,14 @@ def _get_variable_metrics(self, gen_data):
     @torch.no_grad()
     def record_batch(
         self,
+        *,
+        loss: torch.Tensor = torch.tensor(np.nan),
```
Comment (jder) [inline src/ocean_emulators/aggregator/validate/reduced.py:105]: To match the other signatures

File: src/ocean_emulators/aggregator/validate/sub_aggregator.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+        self,
+        *,
+        loss: torch.Tensor,
+        target_data,
+        gen_data,
+        input_data,
+        target_data_norm,
+        gen_data_norm,
+        input_data_norm,
+    ): ...
```
Comment (jder) [inline src/ocean_emulators/aggregator/validate/sub_aggregator.py:23]: Created this abstract class so we can reason over the dict of these things in `ValidateAggregator`

File: .vscode/settings.json (hunk 6 above)
Comment (suryadheeshjith) [reply .vscode/settings.json]: We could include a sample settings.json file on CONTRIBUTING.md?

File: src/ocean_emulators/aggregator/inference/main.py (hunk 7 above)
Comment (suryadheeshjith) [reply src/ocean_emulators/aggregator/inference/main.py:245]: Yes!

File: src/ocean_emulators/aggregator/validate/main.py (hunk 8 above)
Comment (suryadheeshjith) [reply src/ocean_emulators/aggregator/validate/main.py:47]: Cool!

File: src/ocean_emulators/models/base.py [hunk 13]
```diff
@@ -40,7 +40,9 @@ def forward(
             if step == 0:
                 input_tensor = train_data.get_initial_input()
             else:
-                input_tensor = train_data.merge_prognostic_and_boundary(
+                # TODO(jder): this function seems to be unused, resolve
+                # as part of https://github.com/suryadheeshjith/Ocean_Emulator/issues/51
+                input_tensor = train_data.merge_prognostic_and_boundary(  # type: ignore[attr-defined]
```
Comment (jder) [inline src/ocean_emulators/models/base.py:45]: This method doesn't exist on train_data, so silencing this warning for now, unless I misunderstood something? @suryadheeshjith

File: src/ocean_emulators/aggregator/validate/main.py (hunk 9 above)
Comment (suryadheeshjith) [reply src/ocean_emulators/aggregator/validate/main.py:119]: Ah yeah, I did not test this function and it has never been used

Comment (suryadheeshjith) [reply src/ocean_emulators/aggregator/validate/main.py:119]: Super useful later on though

File: src/ocean_emulators/utils/wandb.py [hunk 14]
```diff
@@ -210,26 +218,23 @@ def log_inference_metrics(
...
-                predictions,
-                targets,
-                targets_unnormalized,
-                model_pred_unnormalized,
-            ]
+        if (
+            predictions is not None
+            and targets is not None
+            and targets_unnormalized is not None
+            and model_pred_unnormalized is not None
```
Comment (jder) [inline src/ocean_emulators/utils/wandb.py:225]: This is sad but the type checker didn't understand the previous form meant these things were non-none

File: src/ocean_emulators/utils/wandb.py [hunk 15]
```diff
@@ -210,26 +218,23 @@ def log_inference_metrics(
...
         ):
             for i, var in enumerate(outputs):
                 fig = plt.figure(figsize=(10, 5))
                 plt.plot(
                     range(targets.shape[0]),
-                    targets_unnormalized[:, :, :, i].mean(axis=(1, 2)),
+                    targets_unnormalized[:, :, :, i].mean(dim=(1, 2)),
```
Comment (jder) [inline src/ocean_emulators/utils/wandb.py:231]: TIL that pytorch silently supports both of these but unfortunately the typing declarations only show dim, not axis. AFAICT they are equivalent but this is undocumented.

**Comment by jder** [comment]
ok, I think I fixed and/or silenced all the rest of the new mypy errors, so this is ready for a final review. 🙇

**Comment by suryadheeshjith** [comment]
We definitely need that regression test before merging in this PR. I hope that's okay!

File: .vscode/settings.json (hunk 6 above)
Comment (jder) [reply .vscode/settings.json]: Done, thanks!

**Comment by jder** [comment]
> We definitely need that regression test before merging in this PR. I hope that's okay!

Seems reasonable to me. Let's discuss what it should look like in slack?

File: src/ocean_emulators/models/base.py (hunk 13 above)
Comment (suryadheeshjith) [reply src/ocean_emulators/models/base.py:45]: This [PR](https://github.com/suryadheeshjith/Ocean_Emulator/commit/9bf9391bb918a926b0d2c530fde14fae0d78cfe2) removed this function, very confused how this slipped past all the train eval tests.

Comment (jder) [reply src/ocean_emulators/models/base.py:45]: AFAICT it's because this `forward` function is unused -- we always use `inference` which calls `forward_once` or the stepper, which does the same thing.

Comment (suryadheeshjith) [reply src/ocean_emulators/models/base.py:45]: Ah, we need a test to include forward() then

File: .github/workflows/pre-commit.yml (hunk 5 above)
Comment (alxmrs) [reply .github/workflows/pre-commit.yml:20]: Cool, good to know.

File: .gitmodules [hunk 16]
```diff
@@ -1,3 +0,0 @@
-[submodule "ocean_emulators_main"]
```
Comment (alxmrs) [inline .gitmodules:1]: Hell yeah.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (alxmrs) [reply .pre-commit-config.yaml:28]: Great work!

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -154,6 +154,18 @@ git push --force-with-lease
 
 10. Celebrating submitting your patch to Ocean Emulator — well done!
 
+## VS Code Integration
```
Comment (alxmrs) [inline CONTRIBUTING.md:157]: Thanks for adding this!

File: src/ocean_emulators/aggregator/inference/main.py (hunk 3 above)
Comment (alxmrs) [reply src/ocean_emulators/aggregator/inference/main.py:8]: 100% agree -- I think these package focused imports are more stable than relative paths. And, I'm happy to change these over time.

File: src/ocean_emulators/aggregator/validate/reduced.py (hunk 11 above)
Comment (alxmrs) [reply src/ocean_emulators/aggregator/validate/reduced.py:105]: Wow, I'm thrilled that we have mypy checks now!!

File: src/ocean_emulators/models/base.py (hunk 13 above)
Comment (alxmrs) [reply src/ocean_emulators/models/base.py:45]: Sorry, I think I caused this issue!

**Review by alxmrs — APPROVED** [review]
LGTM! These look like a heroic number of type fixes. Great work.

File: src/ocean_emulators/models/base.py (hunk 13 above)
Comment (jder) [reply src/ocean_emulators/models/base.py:45]: Hopefully fixed in #129

**Comment by jder** [comment]
Discussed offline and decided to merge!

**Comment by alxmrs** [comment]
Wohoo!
